django/template/loaders/cached.py

```diff
@@ -16,15 +16,86 @@ class Loader(BaseLoader):
     def __init__(self, engine, loaders):
         self.template_cache = {}
         self.find_template_cache = {}
+        self.get_template_cache = {}
         self.loaders = engine.get_template_loaders(loaders)
         super(Loader, self).__init__(engine)
 
-    def cache_key(self, template_name, template_dirs):
+    def get_contents(self, origin):
+        return origin.loader.get_contents(origin)
+
+    def get_template(self, template_name, template_dirs=None, skip=None):
+        key = self.cache_key(template_name, template_dirs, skip)
+
+        cached = self.get_template_cache.get(key)
+        if cached:
+            if isinstance(cached, TemplateDoesNotExist):
+                raise cached
+            return cached
+
+        try:
+            template = super(Loader, self).get_template(
+                template_name, template_dirs, skip,
+            )
+        except TemplateDoesNotExist as e:
+            self.get_template_cache[key] = e
+            raise e
```

Comment:
Unless you did this on purpose, a simple `raise` is better, as it avoids a needless frame in the traceback.